akka-cluster/src/main/scala/akka/cluster/CrossDcClusterHeartbeat.scala

```diff
@@ -0,0 +1,365 @@
+
+    // slight optimisation, to avoid searching over all members:
+    state.get(dc) match {
+      case Some(dcMembers) ⇒ dcMembers.exists(_.uniqueAddress == memberUniqueAddress)
+      case _               ⇒ false // since we don't have any entry for this DC, no member will match
+    }
+  }
+
+  /**
+   * Decides if `self` node should become active and monitor other nodes with heartbeats.
+   * Only the `nrOfMonitoredNodesPerDc`-oldest nodes in each DC fulfil this role.
+   */
+  def shouldActivelyMonitorNodes(selfDc: ClusterSettings.DataCenter, selfAddress: UniqueAddress, log: String ⇒ Unit): Boolean = {
+    /** Since we need ordering of oldests guaranteed, we must only look at Up (or Leaving, Exiting...) nodes */
+    def atLeastUp(m: Member): Boolean =
+      m.status != MemberStatus.WeaklyUp && m.status != MemberStatus.Joining
+
+    val selfDcNeighbours: immutable.SortedSet[Member] = state.getOrElse(
+      selfDc,
+      throw new IllegalStateException(s"Self DataCenter was [${selfDc}] however no such entries in ${getClass} present; This is a bug."))
+
+    log("selfDcNeighbours = " + selfDcNeighbours)
+    val selfDcOldOnes = selfDcNeighbours.filter(atLeastUp).take(nrOfMonitoredNodesPerDc)
+    log(s"selfDcOldOnes = ${selfDcOldOnes}")
+
+    // if this node is part of the "n oldest nodes" it should indeed monitor other nodes:
+    val shouldMonitorActively = selfDcOldOnes.exists(_.uniqueAddress == selfAddress)
+    log(s"shouldMonitorActively = ${shouldMonitorActively}")
+    shouldMonitorActively
+  }
+
+  def addMember(m: Member): CrossDcHeartbeatingState = {
+    val dc = m.dataCenter
+    def addMember0(member: Member): CrossDcHeartbeatingState = {
+      val updatedMembers = state.getOrElse(dc, emptyMembersSortedSet) + m
+      this.copy(state = state.updated(dc, updatedMembers)
+      )
+    }
+
+    // we need to remove the member first, to avoid having "duplicates"
+    // this is because the removal and uniqueness we need is only by uniqueAddress
+    // which is not the equals defined on Member; so a "replace" (which this op might be),
+    // must be a "remove then add".
+    removeMember(m)
+    addMember0(m)
+
+  }
+
+  def removeMember(m: Member): CrossDcHeartbeatingState = {
+    val dc = m.dataCenter
+    state.get(dc) match {
+      case Some(dcMembers) ⇒
+        val updatedMembers = dcMembers.filterNot(_.uniqueAddress == m.uniqueAddress)
+        copy(state = state.updated(dc, updatedMembers))
+      case None ⇒
+        this // no change needed, was certainly not present (not even its DC was) 
+    }
+  }
+
+  def activeReceivers: Iterable[UniqueAddress] =
+    for {
+      dc ← state.keys
+      member ← state(dc).take(nrOfMonitoredNodesPerDc)
+    } yield member.uniqueAddress
+
+  def allMembers: Iterable[Member] =
+    state.values.flatMap(ConstantFun.scalaIdentityFunction)
+
+  def heartbeatRsp(from: UniqueAddress): CrossDcHeartbeatingState =
+    if (contains(from)) {
```

Comment:
this should only accept rsp for those that we are actively monitoring now, otherwise it could reintroduce an entry in the failure detector after we have removed it (stopped monitoring it)